packages/astro/src/env/vite-plugin-env.ts

```diff
@@ -0,0 +1,226 @@
+import { loadEnv, type Plugin } from 'vite';
+import type { AstroSettings } from '../@types/astro.js';
+import {
+	ENV_TYPES_FILE,
+	MODULE_TEMPLATE_URL,
+	TYPES_TEMPLATE_URL,
+	VIRTUAL_MODULES_IDS,
+	VIRTUAL_MODULE_SETUP_ID,
+} from './constants.js';
+import type { EnvSchema } from './schema.js';
+import { getType, validateEnvVariable } from './validators.js';
+import { AstroError, AstroErrorData } from '../core/errors/index.js';
+import type fsMod from 'node:fs';
+import { fileURLToPath } from 'node:url';
+
+// TODO: reminders for when astro:env comes out of experimental
+// Types should always be generated (like in types/content.d.ts). That means the client module will be empty
+// and server will only contain getSecret for unknown variables. Then, specifying a schema should only add
+// variables as needed. For secret variables, it will only require specifying SecretValues and it should get
+// merged with the static types/content.d.ts
+
+interface AstroEnvVirtualModPluginParams {
+	settings: AstroSettings;
+	mode: 'dev' | 'build' | string;
+	fs: typeof fsMod;
+}
+
+export function astroEnv({
+	settings,
+	mode,
+	fs,
+}: AstroEnvVirtualModPluginParams): Plugin | undefined {
+	if (!settings.config.experimental.env) {
+		return;
+	}
+	const schema = settings.config.experimental.env.schema ?? {};
+
+	let templates: { client: string; server: string; internal: string } | null = null;
+
+	return {
+		name: 'astro-env-virtual-mod-plugin',
+		enforce: 'pre',
+		buildStart() {
+			const loadedEnv = loadEnv(
+				mode === 'dev' ? 'development' : 'production',
+				fileURLToPath(settings.config.root),
+				''
+			);
+			const validatedVariables = validatePublicVariables({ schema, loadedEnv });
+
+			const clientTemplates = getClientTemplates({ validatedVariables });
+			const serverTemplates = getServerTemplates({ validatedVariables, schema, fs });
+
+			templates = {
+				client: clientTemplates.module,
+				server: serverTemplates.module,
+				internal: `export const schema = ${JSON.stringify(schema)};`,
+			};
+			generateDts({
+				settings,
+				fs,
+				content: getDts({
+					fs,
+					clientPublic: clientTemplates.types,
+					serverPublic: serverTemplates.types.public,
+					serverSecret: serverTemplates.types.secret,
+				}),
+			});
+		},
+		buildEnd() {
+			templates = null;
+		},
+		resolveId(id) {
+			for (const moduleId of Object.values(VIRTUAL_MODULES_IDS)) {
```

Comment:
This could be optimized, probably should given how often this code is called. In the constants module:

```js
export const VIRTUAL_MODULE_ID_VALUES = new Set(Object.values(VIRTUAL_MODULES_IDS));
```

Then check with `VIRTUAL_MODULE_ID_VALUES.has(id)`, which is O(1).